internal/auth/issuer/pam/pam_issuer.go

```diff
@@ -284,36 +314,54 @@ func (s *PAMOIDCProvider) handleAuthorizationCodeGrant(ctx context.Context, req
 		// Confidential client - require client_secret_post authentication
 		if req.ClientSecret == nil {
 			s.log.Warnf("handleAuthorizationCodeGrant: missing client secret")
-			return &pamapi.TokenResponse{Error: lo.ToPtr(ErrorInvalidClient)}, nil
+			return nil, &pamapi.OAuth2Error{
+				Code:             pamapi.InvalidClient,
+				ErrorDescription: lo.ToPtr("Client authentication failed: missing client_secret"),
+			}
 		}
 		if *req.ClientSecret != s.config.ClientSecret {
 			s.log.Warnf("handleAuthorizationCodeGrant: invalid client secret")
-			return &pamapi.TokenResponse{Error: lo.ToPtr(ErrorInvalidClient)}, nil
+			return nil, &pamapi.OAuth2Error{
+				Code:             pamapi.InvalidClient,
+				ErrorDescription: lo.ToPtr("Client authentication failed: invalid client_secret"),
+			}
 		}
 	}
 
 	// Validate and retrieve authorization code
 	codeData, exists := s.codeStore.GetCode(*req.Code)
 	if !exists {
 		s.log.Warnf("handleAuthorizationCodeGrant: authorization code not found")
-		return &pamapi.TokenResponse{Error: lo.ToPtr(ErrorInvalidGrant)}, nil
+		return nil, &pamapi.OAuth2Error{
+			Code:             pamapi.InvalidGrant,
+			ErrorDescription: lo.ToPtr("Authorization code is invalid, expired, or already used"),
+		}
 	}
 
 	// Validate that the client ID matches the stored code
 	if codeData.ClientID != *req.ClientId {
 		s.log.Warnf("handleAuthorizationCodeGrant: invalid client ID - expected=%s, got=%s", codeData.ClientID, *req.ClientId)
-		return &pamapi.TokenResponse{Error: lo.ToPtr(ErrorInvalidGrant)}, nil
+		return nil, &pamapi.OAuth2Error{
+			Code:             pamapi.InvalidGrant,
+			ErrorDescription: lo.ToPtr("Authorization code was not issued to this client"),
+		}
 	}
 
 	// SECURITY: Validate redirect_uri matches the one from authorization request (RFC 6749 Section 4.1.3)
 	// This prevents authorization code interception attacks
 	if req.RedirectUri == nil || *req.RedirectUri == "" {
 		s.log.Warnf("handleAuthorizationCodeGrant: missing redirect_uri")
-		return &pamapi.TokenResponse{Error: lo.ToPtr(ErrorInvalidRequest)}, nil
+		return nil, &pamapi.OAuth2Error{
+			Code:             pamapi.InvalidRequest,
+			ErrorDescription: lo.ToPtr("Missing required parameter: redirect_uri"),
+		}
 	}
 	if codeData.RedirectURI != *req.RedirectUri {
 		s.log.Warnf("handleAuthorizationCodeGrant: redirect_uri mismatch - expected=%s, got=%s", codeData.RedirectURI, *req.RedirectUri)
-		return &pamapi.TokenResponse{Error: lo.ToPtr(ErrorInvalidGrant)}, nil
+		return nil, &pamapi.OAuth2Error{
+			Code:             pamapi.InvalidGrant,
+			ErrorDescription: lo.ToPtr("redirect_uri does not match the one used in the authorization request"),
```

Comment:
IIUC, this error throws when the redirect URI used by the client does not match the one the PAM issuer was configured with.

Maybe the message could be clearer for that case? This is the error I had.